src/java/org/jivesoftware/openfire/plugin/rest/controller/SystemController.java

```diff
@@ -0,0 +1,273 @@
+     *
+     * @return true if at least one system property requires a restart, otherwise false.
+     */
+    public boolean hasSystemPropertyRequiringRestart() {
+        try {
+            SystemProperty.getProperties().stream().filter(SystemProperty::isRestartRequired).forEach(p -> LOG.info("Requires restart: {}. Current value: {}. Initial value: {}", p.getKey(), p.getValue(), p.getDisplayValue()));
+            return SystemProperty.getProperties().stream()
+                .filter(systemProperty -> !systemProperty.getKey().equals("xmpp.domain")) // xmpp.domain can report a false positive. See OF-2399
+                .anyMatch(SystemProperty::isRestartRequired);
+        } catch (Throwable t) {
+            LOG.warn("Unable to determine if there are any system properties that require a restart.", t);
+            return false;
+        }
+    }
+
+    /**
+     * Determine if the core Openfire service is started.
+     *
+     * @return true if Openfire has started, otherwise false.
+     */
+    public boolean isStarted() {
+        return XMPPServer.getInstance().isStarted();
+    }
+
+    /**
+     * Determine if clustering has been fully started, if clustering is enabled.
+     *
+     * When clustering is not enabled, this method will always return 'true'.
+     *
+     * @return true if Openfire has started, otherwise false.
+     */
+    public boolean hasClusteringStartedWhenEnabled() {
+        if (!ClusterManager.isClusteringEnabled()) {
+            return true;
+        }
+        return ClusterManager.isClusteringStarted();
+    }
+
+    /**
+     * Checks if the plugin manager has finished starting all plugins that were available at boot-time.
+     *
+     * Note that the return value does not indicate that plugins succeeded or failed to start: a 'true' return value
+     * indicates only that the plugin manager finished its initial attempt to start all plugins.
+     *
+     * @return True when all (initial) plugins have been started.
+     */
+    public boolean hasPluginManagerExecuted() {
+        return XMPPServer.getInstance().getPluginManager().isExecuted();
+    }
+
+    /**
+     * Verifies that a connection listener for the provided type and encryption level is disabled, or ready to accept connections.
+     *
+     * @param connectionType The type of connection for which to check state
+     * @param encrypted true when the direct-TLS encrypted variant of the listeners is to be checked, otherwise false.
+     * @return True when the connection listener is ready to accept connections, or is disabled.
+     */
+    public boolean isConnectionListenerStartedWhenEnabled(@Nonnull final ConnectionType connectionType, final boolean encrypted) {
+        switch (connectionType) {
+            case BOSH_C2S:
+                if (encrypted) {
+                    return !HttpBindManager.HTTP_BIND_ENABLED.getValue() || HttpBindManager.HTTP_BIND_SECURE_PORT.getValue() <= 0 || HttpBindManager.getInstance().isHttpsBindActive();
```

Comment:
```suggestion
                    return HttpBindManager.HTTP_BIND_ENABLED.getValue() && HttpBindManager.HTTP_BIND_SECURE_PORT.getValue() >= 0 && HttpBindManager.getInstance().isHttpsBindActive();
```

Should it be "If ALL these things are on / have good values"?

Reply:
I don't think so: if the connection type is disabled (either 'enabled == false' or the port number <= 0) then the check should always pass (otherwise, Openfire would never be 'ready').